golang/cmd/event-ingest/main.go

```diff
@@ -0,0 +1,135 @@
+	InsertedAt     time.Time              `json:"inserted_at,omitempty"`
+}
+
+func main() {
+	seeds := []string{"localhost:9092"}
+	ctx := context.Background()
+
+	// Setup kafka consumer
+	cl, err := kgo.NewClient(
+		kgo.SeedBrokers(seeds...),
+		kgo.ConsumerGroup("default"),
+		kgo.ConsumeTopics("test-topic"),
+		kgo.DisableAutoCommit(),
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer cl.Close()
+
+	//setup db connection
+	db, err := sql.Open("postgres", "postgres://user:password@localhost/dbname?sslmode=disable")
+	if err != nil {
+		panic(err)
+	}
+	defer db.Close()
+
+	// Set up prepared statement for insert
+	stmt, err := db.Prepare(`INSERT INTO events (organization_id, customer_id, cust_id, event_name, time_created, properties, idempotency_id, inserted_at)
+		VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
+		ON CONFLICT DO NOTHING`)
+	if err != nil {
+		panic(err)
+	}
+
+	for {
+		fetches := cl.PollFetches(ctx)
+		if fetches == nil {
+			continue
+		}
+		if fetches.IsClientClosed() {
+			return
+		}
+		if errs := fetches.Errors(); len(errs) > 0 {
+			// All errors are retried internally when fetching, but non-retriable errors are
+			// returned from polls so that users can notice and take action.
+			panic(fmt.Sprint(errs))
+		}
+
+		events := make([]*Event, 0)
+		fetches.EachRecord(func(r *kgo.Record) {
+			event := &Event{}
+			err := json.Unmarshal(r.Value, event)
+			if err != nil {
+				fmt.Printf("Error unmarshalling event: %s\n", err)
+				return
+			}
+			event.CustomerID = nil
+			event.InsertedAt = time.Now()
+			events = append(events, event)
+
+			if len(events) >= batchSize {
+				err = insertEvents(stmt, events, db, ctx)
+				if err != nil {
+					fmt.Printf("Error inserting events into database: %s\n", err)
+				}
+				events = make([]*Event, 0)
+				if err := cl.CommitUncommittedOffsets(context.Background()); err != nil {
+					fmt.Printf("commit records failed: %v", err)
+				}
+			}
+		})
+
+		if len(events) > 0 {
+			err := insertEvents(stmt, events, db, ctx)
+			if err != nil {
+				fmt.Printf("Error inserting events into database: %s\n", err)
+				return
+			}
+			if err := cl.CommitUncommittedOffsets(context.Background()); err != nil {
+				fmt.Printf("commit records failed: %v", err)
+			}
+		}
+	}
+}
+
+func insertEvents(stmt *sql.Stmt, events []*Event, db *sql.DB, ctx context.Context) error {
+	tx, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("error starting transaction: %s", err)
+	}
+	defer tx.Rollback()
+
+	// Create a slice to hold the values to be inserted
+	var values []interface{}
+
+	// Loop through the events and add their values to the slice
+	for _, event := range events {
+		values = append(values, event.OrganizationID, event.CustomerID, event.CustID, event.EventName, event.TimeCreated, event.Properties, event.IdempotencyID, event.InsertedAt)
+	}
+
+	// Execute the batch insert statement with the values slice
+	_, err = stmt.Exec(values...)
```

Comment:
It would surprise me if this works, unless I'm missing something. AFAIK you can't pass multiple copies of arbitrary values to the insert. It feels like this code is held over from the Copy code.

If you do `insert into x (a) values ($1)` then you can only pass one argument to `Exec` (being the value of `a`).